### DISCUSSION TITLE
[A-L] (23 luglio 2024) Esercizio 1

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/progettazione-di-algoritmi#110] [Esami A-L]
**Esercizio 1:** un grafo non-diretto $G$ è detto _equanime_ se per ogni vertice $x$ , ogni albero di visita BFS $T_1$ con radice $x$ e ogni albero di visita DFS $T_2$ con radice $x$ , abbiamo che $T_1 = T_2$ . Dare lo pseudocodice di un algoritmo $O(n)$ per stabilire se un grafo dato in input è equanime.

**Comment by Elia-Belli** [comment]
$$T_1 = T_2 \iff \text{le componenti di $G$ sono alberi}$$

Un albero è un grafo connesso aciclico, e quindi senza archi all'indietro: possiamo effettuare un DFS su $G$ terminando la visita se troviamo un arco all'indietro, se $G$ ne possiede uno la DFS lo troverà obbligatoriamente entro $n$ iterazioni.

```python
def equanime(G):
    P = [-1]*n # vettore dei padri DFS

    for v in G.V:  # G non è detto sia connesso, allora eseguo il DFS su tutte le componenti
        if(P[v] == -1):
            res = DFS_eq(G,v,P)
            if(res == False):
                return res
    return True

def DFS_eq(G,v,P):
    P[v] = v 
    S = [v] # stack DFS
    
    while(S != []):
        v = S.top()
        for w in v.adiacenti:
            if(P[w] == -1):
                P[w] = v
                S.append(w)
            elif(w != P[v]):     # (condizione per evitare di considerare il padre di v come adiacente) 
                return False     # se ho già visitato w allora c'è un arco all'indietro
    
    return True
```

**Reply by cek753** [reply]
L'algortimo è corretto, credo solo ci sia un refuso poichè al posto di s.pop() hai messo s.top(), ovvero guardi l'ultimo elemento quando in realtà andrebbe rimosso

**Comment by Near047** [comment]
Allora forse sto sbagliando ma non capisco come non possa essere così.
L'unico modo che si ha affinchè un BFS e un DFS ritronino lo stesso T è avere E(G) = V(G)-1.

Poi anche se non è connesso ma li dipende molto da dove si inizia, ma contiamo che un algoritmo prima controlla se il grafo è connesso e poi calcola se E(G) = V(G)-1, se sì allora significa che per ogni vertice ci sta solo un arco entrante tranne per la radice, aka impossibile che bfs e dfs differiscano. 

Se qualcuno mi vuole correggere, vi prego.
